inlong-sdk/sort-sdk/src/main/java/org/apache/inlong/sdk/sort/impl/tube/InLongTubeFetcherImpl.java

```diff
@@ -0,0 +1,291 @@
+
+import com.google.common.base.Splitter;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeSet;
+import java.util.concurrent.TimeUnit;
+import org.apache.inlong.sdk.sort.api.ClientContext;
+import org.apache.inlong.sdk.sort.api.InLongTopicFetcher;
+import org.apache.inlong.sdk.sort.api.SysConstants;
+import org.apache.inlong.sdk.sort.entity.InLongMessage;
+import org.apache.inlong.sdk.sort.entity.InLongTopic;
+import org.apache.inlong.sdk.sort.entity.MessageRecord;
+import org.apache.inlong.sdk.sort.util.StringUtil;
+import org.apache.inlong.tubemq.client.config.ConsumerConfig;
+import org.apache.inlong.tubemq.client.config.TubeClientConfig;
+import org.apache.inlong.tubemq.client.consumer.ConsumerResult;
+import org.apache.inlong.tubemq.client.consumer.PullMessageConsumer;
+import org.apache.inlong.tubemq.corebase.Message;
+import org.apache.inlong.tubemq.corebase.TErrCodeConstants;
+import org.apache.pulsar.shade.org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InLongTubeFetcherImpl extends InLongTopicFetcher {
+
+    private static final Logger LOG = LoggerFactory.getLogger(InLongTubeFetcherImpl.class);
+    private PullMessageConsumer messageConsumer;
+
+    private volatile boolean stopConsume = false;
+    private volatile Thread fetchThread;
+    private long sleepTime = 0L;
+    private int emptyPollTimes = 0;
+
+    public InLongTubeFetcherImpl(InLongTopic inLongTopic, ClientContext context) {
+        super(inLongTopic, context);
+    }
+
+    @Override
+    public boolean init(Object object) {
+        TubeConsumerCreater tubeConsumerCreater = (TubeConsumerCreater) object;
+        TubeClientConfig tubeClientConfig = tubeConsumerCreater.getTubeClientConfig();
+        try {
+            ConsumerConfig consumerConfig = new ConsumerConfig(tubeClientConfig.getMasterInfo(),
+                    context.getConfig().getSortTaskId());
+
+            messageConsumer = tubeConsumerCreater.getMessageSessionFactory().createPullConsumer(consumerConfig);
+            if (messageConsumer != null) {
+                TreeSet<String> filters = null;
+                if (inLongTopic.getProperties() != null && inLongTopic.getProperties().containsKey(
+                        SysConstants.TUBE_TOPIC_FILTER_KEY)) {
+                    filters = (TreeSet<String>) inLongTopic.getProperties().get(SysConstants.TUBE_TOPIC_FILTER_KEY);
+
+                }
+                messageConsumer.subscribe(inLongTopic.getTopic(), filters);
+                messageConsumer.completeSubscribe();
+
+                String threadName = "sort_sdk_fetch_thread_" + StringUtil
+                        .formatDate(new Date(), "yyyy-MM-dd HH:mm:ss.SSS");
+                this.fetchThread = new Thread(new Fetcher(), threadName);
+                this.fetchThread.start();
```

Comment:
fetch Thread can use thread pool.
One topic has one thread, the thread is too much.

Reply:
Creating threads according to topic level will not lead to the problem of too many threads. On the contrary, it can simplify the program logic design. This has been verified in other ways of creating threads based on partition. If you use thread pool, it will also bring some other problems in use